coderd/database/modelmethods.go

```diff
@@ -235,6 +241,199 @@ func (s APIKeyScopes) Name() rbac.RoleIdentifier {
+	// Empty DB list means no additional restriction.
+	if len(dbList) == 0 {
+		// Defensive: API keys should always persist a wildcard allow list, but
+		// treating an empty slice as "no additional filter" preserves legacy data.
+		return scopeList
+	}
+
+	// If scope already allows everything, the db list is authoritative.
+	scopeAll := allowListAll(scopeList)
+	dbAll := dbList.all()
+
+	switch {
+	case scopeAll && dbAll:
+		return []rbac.AllowListElement{rbac.AllowListAll()}
+	case scopeAll:
+		return normalizeDBAllowList(dbList)
+	case dbAll:
+		return scopeList
+	}
+
+	// Otherwise compute intersection.
+	resultSet := make(map[string]rbac.AllowListElement)
+	for _, scopeElem := range scopeList {
+		matching := intersectAllow(scopeElem, dbList)
+		for _, elem := range matching {
+			resultSet[elem.String()] = elem
+		}
+	}
+
+	if len(resultSet) == 0 {
+		return []rbac.AllowListElement{}
+	}
+
+	result := make([]rbac.AllowListElement, 0, len(resultSet))
+	for _, elem := range resultSet {
+		result = append(result, elem)
+	}
+
+	slices.SortFunc(result, func(a, b rbac.AllowListElement) int {
+		if a.Type == b.Type {
+			return strings.Compare(a.ID, b.ID)
+		}
+		return strings.Compare(a.Type, b.Type)
+	})
+	return result
+}
+
+func allowListAll(elements []rbac.AllowListElement) bool {
+	if len(elements) == 0 {
+		return false
+	}
+	for _, e := range elements {
+		if e.Type == policy.WildcardSymbol && e.ID == policy.WildcardSymbol {
+			return true
+		}
+	}
+	return false
+}
+
+func (l AllowList) all() bool {
+	if len(l) == 0 {
+		return false
+	}
+	for _, e := range l {
+		if e.Type == policy.WildcardSymbol && e.ID == policy.WildcardSymbol {
+			return true
+		}
+	}
+	return false
+}
+
+func normalizeDBAllowList(dbList AllowList) []rbac.AllowListElement {
+	if len(dbList) == 0 {
+		return []rbac.AllowListElement{}
+	}
+	out := make([]rbac.AllowListElement, 0, len(dbList))
+	for _, e := range dbList {
+		out = append(out, rbac.AllowListElement{Type: e.Type, ID: e.ID})
+	}
+	return out
+}
+
+// intersectAllow returns the set of permit entries that satisfy both the scope
+// element and the database allow list.
+func intersectAllow(scopeElem rbac.AllowListElement, dbList AllowList) []rbac.AllowListElement {
+	// Scope element is wildcard -> intersection is db list.
+	if scopeElem.Type == policy.WildcardSymbol && scopeElem.ID == policy.WildcardSymbol {
+		return normalizeDBAllowList(dbList)
+	}
+
+	result := make([]rbac.AllowListElement, 0)
+	for _, dbElem := range dbList {
+		// DB entry wildcard -> keep scope element.
+		if dbElem.Type == policy.WildcardSymbol && dbElem.ID == policy.WildcardSymbol {
+			result = append(result, scopeElem)
+			continue
+		}
+
+		if !typeMatches(scopeElem.Type, dbElem.Type) {
+			continue
+		}
+
+		idMatch := scopeElem.ID == policy.WildcardSymbol || dbElem.ID == policy.WildcardSymbol || scopeElem.ID == dbElem.ID
+		if !idMatch {
+			continue
+		}
```

Comment:
You made `typeMatches` a function, but not the idmatch. 

Nit for consistency, do 1 or the other